src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -927,13 +927,13 @@ internal void FinishOperationSyncSuccess(int bytesTransferred, SocketFlags flags
                     {
                         try
                         {
-                            if (_remoteEndPoint!.AddressFamily == _socketAddress!.Family)
+                            if (_remoteEndPoint!.AddressFamily == AddressFamily.InterNetworkV6 && _socketAddress!.Family == AddressFamily.InterNetwork)
                             {
-                                _remoteEndPoint = _remoteEndPoint!.Create(_socketAddress);
+                                _remoteEndPoint = new IPEndPoint(_socketAddress.GetIPAddress().MapToIPv6(), _socketAddress.GetPort());
                             }
```

Comment:
Is this something IPEndPoint.Create(SocketAddress) should be doing instead of doing it at this one call site?

Reply:
this one is tricky - we have cases when the address is IPv4 and `Create` would just work - but we want IPv6 IPv4Mapped addresses for DualMode sockets.  I guess mostly for compatibility reasons so this is handled as special case.